akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala

```diff
@@ -617,19 +621,45 @@ private[remote] class ArteryTransport(_system: ExtendedActorSystem, _provider: R
 
   private def runInboundOrdinaryMessagesStream(compression: InboundCompressions): Unit = {
     val completed =
-      if (remoteSettings.TestMode) {
-        val (mgmt, c) = aeronSource(ordinaryStreamId, envelopePool)
-          .via(inboundFlow(compression))
-          .viaMat(inboundTestFlow)(Keep.right)
-          .toMat(inboundSink)(Keep.both)
-          .run()(materializer)
-        testStages.add(mgmt)
-        c
+      if (inboundLanes == 1) {
+        if (remoteSettings.TestMode) {
+          val (mgmt, c) = aeronSource(ordinaryStreamId, envelopeBufferPool)
+            .via(inboundFlow(compression))
+            .viaMat(inboundTestFlow)(Keep.right)
+            .toMat(inboundSink(envelopeBufferPool))(Keep.both)
+            .run()(materializer)
+          testStages.add(mgmt)
+          c
+        } else {
+          aeronSource(ordinaryStreamId, envelopeBufferPool)
+            .via(inboundFlow(compression))
+            .toMat(inboundSink(envelopeBufferPool))(Keep.right)
+            .run()(materializer)
+        }
+
       } else {
-        aeronSource(ordinaryStreamId, envelopePool)
+        val source = aeronSource(ordinaryStreamId, envelopeBufferPool)
           .via(inboundFlow(compression))
-          .toMat(inboundSink)(Keep.right)
-          .run()(materializer)
+          .watchTermination()(Keep.right)
+
+        val lane = inboundSink(envelopeBufferPool).async
+        val partitionFun: InboundEnvelope ⇒ Int = {
+          envelope ⇒
+            envelope.recipient match {
+              case OptionVal.Some(r) ⇒ math.abs(r.path.uid) % inboundLanes
+              case OptionVal.None    ⇒ 0
+            }
+        }
+        val sink = Sink.combine(lane, lane, List.fill(inboundLanes - 2)(lane): _*) { n ⇒
+          Partition[InboundEnvelope](n, partitionFun)
+        }
+
+        val completed = source.to(sink).run()(materializer)
+
+        // FIXME completed is the upstream watchTermination, but we need the mat value of the inboundSink for correct error handling
+        // FIXME we also need mat values of the TestMode stage, also part of the lane
```

Comment:
Yes, you can try the filter version. We will need a partition version for many other use-cases, too, so eventually we will have one.